src/graphql/data/TopTokens.ts

```diff
@@ -0,0 +1,222 @@
+        break
+    }
+
+    return sortAscending ? tokenArray.reverse() : tokenArray
+  }, [tokens, sortMethod, sortAscending])
+}
+
+function useFilteredTokens(tokens: PrefetchedTopToken[]) {
+  const filterString = useAtomValue(filterStringAtom)
+  const favorites = useAtomValue(favoritesAtom)
+  const showFavorites = useAtomValue(showFavoritesAtom)
+
+  const lowercaseFilterString = useMemo(() => filterString.toLowerCase(), [filterString])
+
+  return useMemo(() => {
+    if (!tokens) {
+      return []
+    }
+
+    let returnTokens = tokens
+    if (showFavorites) {
+      returnTokens = returnTokens?.filter((token) => (token?.address && favorites.includes(token.address)) ?? false)
+    }
+    if (lowercaseFilterString) {
+      returnTokens = returnTokens?.filter((token) => {
+        const addressIncludesFilterString = token?.address?.toLowerCase().includes(lowercaseFilterString)
+        const nameIncludesFilterString = token?.name?.toLowerCase().includes(lowercaseFilterString)
+        const symbolIncludesFilterString = token?.symbol?.toLowerCase().includes(lowercaseFilterString)
+        return (nameIncludesFilterString || symbolIncludesFilterString || addressIncludesFilterString) ?? false
+      })
+    }
+    return returnTokens
+  }, [tokens, showFavorites, lowercaseFilterString, favorites])
+}
+
+const PAGE_SIZE = 20
+//const MAX_COUNT = 100
+
+function toContractInput(token: PrefetchedTopToken) {
+  return {
+    address: token?.address ?? '',
+    chain: token?.chain ?? 'ETHEREUM',
+  }
+}
+
+export type TopToken = NonNullable<TopTokens_TokensQuery['response']['tokens']>[number]
+export function useTopTokens(prefetchedTokens: TopTokens100Query['response']) {
+  // TODO: add filtering by favorites and filter string
+  const duration = toHistoryDuration(useAtomValue(filterTimeAtom))
+  const environment = useRelayEnvironment()
+  const [tokens, setTokens] = useState<TopToken[]>()
+
+  const [page, setPage] = useState(1)
+  const selectedTokens = useFilteredTokens(useSortedTokens(prefetchedTokens.topTokens))
+  const [isFetching, setIsFetching] = useState(true)
+
+  // TopTokens should ideally be fetched with usePaginationFragment. The backend does not current support graphql cursors;
+  // in the meantime, fetchQuery is used, as other relay hooks do not allow the refreshing and lazy loading we need
+  const loadTokens = useCallback(
+    (contracts: ContractInput[], onSuccess: (data: TopTokens_TokensQuery['response'] | undefined) => void) => {
+      fetchQuery<TopTokens_TokensQuery>(
+        environment,
+        tokensQuery,
+        { contracts, duration },
+        { fetchPolicy: 'store-or-network' }
+      )
+        .toPromise()
+        .then(onSuccess)
+    },
+    [duration, environment]
+  )
+
+  const loadMoreTokens = () => {
+    const contracts = selectedTokens.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE).map(toContractInput)
```

Comment:
is `contracts` the correct name here? Should it be `contractInputs`?

Reply:
I'm following the general naming scheme of queries using $contracts which are of type ContractInput